### ISSUE TITLE
Infinite loop somewhere after type inference

### ISSUE THREAD

**Issue by owickstrom** [oden-lang/oden#73]
When compiling the following program `oden` hangs indefinitely.

```go
package chain/main

import fmt

compose : forall a b c. (b -> c) -> (a -> b) -> a -> c
compose(f, g, x) = f(g(x))

chain : forall a. []{(a -> a)} -> (a -> a)
chain(fs) = if len(fs) == 0 then (x) -> x else compose(fs[0], chain(fs[1:]))

shout(s) = s ++ "!"

x = (chain([]{shout, shout, shout}))("fix meee")

main() = fmt.Println(x)

```

By using the print commands (that are presently only on the `extensible-records` branch) it can be verified that it occurs after inference:

```
$ oden print-inferred src/chain.oden
package chain/main
import fmt
compose : forall a b c. c -> b -> a -> c -> a -> b
compose = (f) -> (g) -> (x) -> f(g(x))
chain : forall a. []{a -> a} -> a -> a
chain = (fs) -> if (len(fs) == 0) then (x) -> x else compose(fs[0])(chain(fs[1 :]))
shout : string -> string
shout = (s) -> (s ++ "!")
x :: string
x = chain([]{shout, shout, shout})("fix meee")
main : -> ()
main = () -> fmt.Println([]{x})
$ oden print-compiled src/chain.oden
...
```

**Comment by kcsongor** [comment]
The issue doesn't seem to happen when chain's type is restricted to
`chain :: forall a. []{(a -> a)} -> (string -> string)`
however, I feel like the latter shouldn't even type check?

**Comment by owickstrom** [comment]
No that seems very strange, will check that out.

**Comment by AlbinTheander** [comment]
Update. In 0.3.0-alpha13, this no longer loops but results in an error message.

Rewriting it to use the `println` function:

```Go
package chain/main

compose : forall a b c. (b -> c) -> (a -> b) -> a -> c
compose(f, g, x) = f(g(x))

chain : forall a. []{(a -> a)} -> (a -> a)
chain(fs) = if len(fs) == 0 then (x) -> x else compose(fs[0], chain(fs[1:]))

shout(s) = s ++ "!"

x = (chain([]{shout, shout, shout}))("fix meee")

main() = println(x)
```

yields the error

```
src/main.oden:7:63: error: Cannot unify types `[]{_t0 -> _t0}` and `[]{_t7 -> _t8}`
```
referring to the recursive call to `chain` towards the end of the line.

**Comment by AlbinTheander** [comment]
It may be something with unification of slices with functions. The following code:

```Go
package count/main

count : forall a. []{a -> a} -> int
count(fs) = len(fs)

shout : string -> string
shout(s) = s ++ "!"

main() = println(count([]{shout, shout, shout}))
```
fails with
```
main.oden:9:18: error: Cannot unify types `[]{_t3 -> _t3}` and `[]{string -> string}`
```

**Comment by AlbinTheander** [comment]
Nope, the last one is a separate issue. Submitted it here: #87.

When it's fixed, we the original example still hangs.

**Comment by AlbinTheander** [comment]
There seems to be something happening in the loop, at least.

When my sublime-plugin tries to lint the code, oden hangs and after a few minutes is using 32Gb of my memory, almost crashing my laptop.

**Comment by owickstrom** [comment]
Solved in 96bac9ce6a27f35001b0c71088e940b92023b377.
